src/Sprite/TilesetLoader.cpp

```diff
@@ -21,6 +21,27 @@ namespace Tileset
 		return tag == TagFileSignature;
 	}
 
+	BitmapFile ReadTileset(Stream::BidirectionalReader& reader)
+	{
+		if (PeekIsCustomTileset(reader)) {
+			return ReadCustomTileset(reader);
+		}
+
+		BitmapFile tileset;
+		
+		try {
+			tileset = BitmapFile::ReadIndexed(reader);
+		}
```

Comment:
This would technically default initialize `tileset`, and then copy assign `tileset`. Moving the declaration into the `try` block would allow for single stage initialization. It would also mean `BitmapFile` wouldn't be required to have a default constructor. That may be a worthy consideration, since it's not immediately clear what kind of usable `BitmapFile` could be generated without any arguments.

I do understand the structure of the remaining code would be affected by such a change, due to scope changes. Though I suppose the rest of the function could be moved into the `try` block. That would affect exception generation from `ValidateTileset`, which would be caught and translated by the `catch` block. Though the `throw` does include the original exception message in the new exception message, so that may be fine.

Reply:
I thought about using a unique_ptr to represent `tileset` so it could be initialized within the try block without a default initialization. I didn't want to deal with a pointer though if I could help it although maybe it isn't that big of a deal.

I also considered what you brought up about just allowing the error message to encapsulate the outer throw. The message will appear a bit duplicative to the end reader, but still clear enough. Maybe this is the easier solution and worth avoiding a pointer?

Reply:
Yeah, I like to avoid pointers when possible. Heap allocations and not nearly as cheap as local stack allocations.

C++ is usually pretty good at providing alternatives to avoid pointers unless they are actually strictly required from a theoretical standpoint. It doesn't normally make sense to use pointers for easier syntax.